src/controller/quota/controller.go

```diff
@@ -167,107 +164,46 @@ func (c *controller) List(ctx context.Context, query *q.Query, options ...Option
+	f := func() error {
+		q, err := c.quotaMgr.GetByRef(ctx, reference, referenceID)
 		if err != nil {
-			return err
+			return lib.NewRetryAbort(err)
 		}
 
 		hardLimits, err := q.GetHard()
 		if err != nil {
-			return err
+			return lib.NewRetryAbort(err)
 		}
 
 		used, err := q.GetUsed()
 		if err != nil {
-			return err
+			return lib.NewRetryAbort(err)
 		}
 
-		reserved, err := c.getReservedResources(ctx, reference, referenceID)
+		newUsed, err := op(hardLimits, used)
 		if err != nil {
-			log.G(ctx).Errorf("failed to get reserved resources for %s %s, error: %v", reference, referenceID, err)
-			return err
+			return lib.NewRetryAbort(err)
 		}
 
-		newReserved := types.Add(reserved, resources)
-
-		if err := quota.IsSafe(hardLimits, types.Add(used, reserved), types.Add(used, newReserved), false); err != nil {
-			return errors.DeniedError(err).WithMessage("Quota exceeded when processing the request of %v", err)
-		}
+		q.SetUsed(newUsed)
 
-		if err := c.setReservedResources(ctx, reference, referenceID, newReserved); err != nil {
-			log.G(ctx).Errorf("failed to set reserved resources for %s %s, error: %v", reference, referenceID, err)
-			return err
+		err = c.quotaMgr.Update(ctx, q)
+		if err != nil && !errors.Is(err, orm.ErrOptimisticLock) {
+			return lib.NewRetryAbort(err)
 		}
 
-		return nil
+		return err
 	}
 
-	return orm.WithTransaction(reserve)(ctx)
-}
-
-func (c *controller) unreserveResources(ctx context.Context, reference, referenceID string, resources types.ResourceList) error {
-	unreserve := func(ctx context.Context) error {
-		if _, err := c.quotaMgr.GetByRefForUpdate(ctx, reference, referenceID); err != nil {
-			return err
-		}
-
-		reserved, err := c.getReservedResources(ctx, reference, referenceID)
-		if err != nil {
-			log.G(ctx).Errorf("failed to get reserved resources for %s %s, error: %v", reference, referenceID, err)
-			return err
-		}
-
-		newReserved := types.Subtract(reserved, resources)
-		// ensure that new used is never negative
-		if negativeUsed := types.IsNegative(newReserved); len(negativeUsed) > 0 {
-			return fmt.Errorf("reserved resources is negative for resource(s): %s", quota.PrettyPrintResourceNames(negativeUsed))
-		}
-
-		if err := c.setReservedResources(ctx, reference, referenceID, newReserved); err != nil {
-			log.G(ctx).Errorf("failed to set reserved resources for %s %s, error: %v", reference, referenceID, err)
-			return err
-		}
-
-		return nil
+	options := []lib.RetryOption{
+		lib.RetryTimeout(defaultRetryTimeout),
+		lib.RetryBackoff(false),
+		lib.RetryCallback(func(err error, sleep time.Duration) {
+			log.G(ctx).Debugf("failed to update the quota usage, error: %v", err)
+		}),
 	}
-
-	return orm.WithTransaction(unreserve)(ctx)
+	return lib.RetryUntil(f, options...)
 }
 
 func (c *controller) Refresh(ctx context.Context, reference, referenceID string, options ...Option) error {
```

Comment:
Since we add retry for the Refresh, it may cause the performance issue for API on DeleteArtifact & DeleteRepository in the failure case, is it expected?

Reply:
Using the `select for update` has the same problem.